interop: add predeploys

0xfuturistic opened the pull request.

0xfuturistic committed:
contracts-bedrock: create CrossL2Inbox

0xfuturistic committed:
contracts-bedrock: refactor CrossL2Inbox

0xfuturistic committed:
contracts-bedrock: move CrossL2Inbox to src/L2

0xfuturistic committed:
contracts-bedrock: create L2ToL2CrossDomainMessenger

0xfuturistic committed:
contracts-bedrock: update L1Block for interop

0xfuturistic committed:
contracts-bedrock: refactor L1Block.t.sol

0xfuturistic committed:
contracts-bedrock: implement initialize method for CrossL2Inbox

0xfuturistic committed:
contracts-bedrock: update snapshots

0xfuturistic committed:
op-bindings: make

0xfuturistic committed:
contracts-bedrock: update semver-lock

0xfuturistic committed:
contracts-bedrock: refactor logic in CrossL2Inbox

0xfuturistic committed:
contracts-bedrock: refactor logic for CrossL2Inbox

0xfuturistic committed:
contracts-bedrock: make CrossL2Inbox not abstract

0xfuturistic committed:
contracts-bedrock: refactor L2ToL2CrossDomainMessenger

0xfuturistic committed:
contracts-bedrock: update L1Block to more closely follow specs

0xfuturistic committed:
contracts-bedrock: add encodeSetL1BlockValuesInterop to encoding lib

0xfuturistic committed:
contracts-bedrock: add interop tests to L1Block

0xfuturistic committed:
op-bindings: make

0xfuturistic committed:
contracts-bedrock: improve calldata extraction for interop in L1Block

0xfuturistic committed:
op-bindings: make

0xfuturistic committed:
contracts-bedrock: add missing doc for Encoding

0xfuturistic committed:
contracts-bedrock: fix logic for setL1BlockValuesInterop

0xfuturistic committed:
contracts-bedrock: refactor logic for setL1BlockValuesInterop

0xfuturistic committed:
op-bindings: make

0xfuturistic committed:
contracts-bedrock: add TODO comment for setL1BlockValuesInterop

0xfuturistic committed:
contracts-bedrock move invariant into setL1BlockValuesInterop

0xfuturistic committed:
contracts-bedrock: refactor tests for L1Block

0xfuturistic committed:
contracts-bedrock: add predeploys CROSS_L2_INBOX, L2_TO_L2_CROSS_DOMAIN_MESSENGER

0xfuturistic committed:
contracts-bedrock: refactor L2ToL2CrossDomainMessenger.t.sol

0xfuturistic committed:
contracts-bedrock: move immutable into storage in L2ToL2CrossDomainMessenger

0xfuturistic committed:
op-chain-ops: add Predeploys CrossL2Inbox, L2ToL2CrossDomainMessenger

0xfuturistic committed:
contracts-bedrock: move l1Block into immutable from storage in CrossL2Inbox

0xfuturistic committed:
contracts-bedrock: move CROSS_L2_INBOX into immutable from storage in L2ToL2CrossDomainMessenger

0xfuturistic committed:
op-chain-ops: update immutable variables for CrossL2Inbox, L2ToL2CrossDomainMessenger

0xfuturistic committed:
contracts-bedrock: update interfaces in Predeploys.sol

0xfuturistic committed:
contracts-bedrock: drop Initializable from CrossL2Inbox

0xfuturistic committed:
contracts-bedrock: update snapshots

0xfuturistic committed:
op-bindings: add CrossL2Inbox, L2ToL2CrossDomainMessenger' artifacts to local

0xfuturistic committed:
op-bindings: make

0xfuturistic committed:
sdk: fix l1Block forge artifact import

0xfuturistic committed:
sdk: add CrossL2Inbox, L2ToL2CrossDomainMessenger to contracts

0xfuturistic committed:
contracts-bedrock: introduce constructor parameter for CrossL2Inbox

0xfuturistic committed:
contracts-bedrock: introduce constructor parameter for L2ToL2CrossDomainMessenger

0xfuturistic committed:
op-chain-ops: fix immutables deployment for CrossL2Inbox, L2ToL2CrossDomainMessenger

0xfuturistic committed:
op-bindings: make

0xfuturistic committed:
op-bindings: drop ProxyDisabled for CrossL2Inbox, L2ToL2CrossDomainMessenger

0xfuturistic committed:
op-chain-ops: drop immutable vars from config for CrossL2Inbox, L2ToL2CrossDomainMessenger

0xfuturistic committed:
op-chain-ops: hardcode predeploys into deploy for CrossL2Inbox, L2ToL2CrossDomainMessenger

0xfuturistic committed:
contracts-bedrock: add CrossL2Inbox, L2ToL2CrossDomainMessenger to artifacts script

0xfuturistic committed:
contracts-bedrock: drop constructor arg for CrosssL2Inbox, L2ToL2CrossDomainMessenger

0xfuturistic committed:
op-bindings: make

0xfuturistic committed:
contracts-bedrock: add tests for CrossL2Inbox

0xfuturistic committed:
contracts-bedrock: add tests for L2ToL2CrossDomainMessenger

0xfuturistic committed:
contracts-bedrock: update semver-lock

tynes commented on packages/contracts-bedrock/src/L2/CrossL2Inbox.sol:
```diff
@@ -0,0 +1,97 @@
+        require(_id.timestamp <= block.timestamp, "Invalid timestamp"); // timestamp invariant
+        uint256 chainId_;
+        assembly {
+            chainId_ := mload(add(_id, 0x80))
```
This seems fragile to changes in the struct layout. Obviously we would want test coverage that would fail such that a struct layout change would be caught, still feels like using assembly here doesn't really give much of a benefit

tynes commented on packages/contracts-bedrock/src/L2/CrossL2Inbox.sol:
```diff
@@ -0,0 +1,97 @@
+    /// @param _msg The message payload, matching the initiating message.
+    /// @param _id A Identifier pointing to the initiating message.
+    /// @param _target Account that is called with _msg.
+    function executeMessage(bytes calldata _msg, Identifier calldata _id, address _target) public payable {
```
The abi doesn't match the spec here, any reason as to why?

tynes commented on packages/contracts-bedrock/src/L2/CrossL2Inbox.sol:
```diff
@@ -0,0 +1,97 @@
+            tstore(CHAINID_SLOT, chainId_)
+        }
+
+        bool success = SafeCall.call({ _target: _target, _gas: gasleft(), _value: msg.value, _calldata: _msg });
```
Can we add another variant of `SafeCall.call` that doesn't take a `gas` and just passes along all gas?

0xfuturistic committed:
contracts-bedrock: update gas-snapshot

0xfuturistic committed:
contracts-bedrock: move L1_BLOCK immutable into storage in CrossL2Inbox

0xfuturistic committed:
contracts-bedrock: move CROSS_L2_INBOX immutable into storage in contracts-bedrock: move L1_BLOCK immutable into storage in CrossL2Inbox

0xfuturistic committed:
op-bindings: make

0xfuturistic committed:
contracts-bedrock: improve error messages in CrossL2Inbox

0xfuturistic committed:
contracts-bedrock: use variables for leading vars in CrossL2Inbox

0xfuturistic committed:
contracts-bedrock: make relayMessage payable for L2ToL2CrossDomainMessenger

0xfuturistic committed:
contracts-bedrock: add events for L2ToL2CrossDomainMessenger

0xfuturistic committed:
contracts-bedrock: make call success not conditional on successful call of L2ToL2CrossDomainMessenger

0xfuturistic committed:
contracts-bedrock: improve error messages for L2ToL2CrossDomainMessenger

0xfuturistic committed:
contracts-bedrock: add new test for CrossL2Inbox + refactor tests

0xfuturistic committed:
contracts-bedrock: refactor tests for L2ToL2CrossDomainMessenger

0xfuturistic committed:
op-bindings: make

0xfuturistic committed:
contracts-bedrock: fix test_executeMessage_unsuccessfullSafeCall_fails

0xfuturistic committed:
contracts-bedrock: fix Initializable.t.sol

0xfuturistic committed:
op-bindings: make

0xfuturistic committed:
contracts-bedrock: drop fallback for L2ToL2CrossDomainMessenger.t.sol

0xfuturistic committed:
contracts-bedrock: update snapshots

0xfuturistic committed:
upgrade slither to version 0.10.1

need to because otherwise it wont recognize tload or tstore (transient storage)

0xfuturistic committed:
op-node: fix fuzz_parsers_test

0xfuturistic committed:
contracts-bedrock: add callWithAllGas to SafeCall library

0xfuturistic committed:
contracts-bedrock: use callWithAllGas in CrossL2Inbox, L2ToL2CrossDomainMessenger

0xfuturistic committed:
contracts-bedrock: fix CrossL2Inbox executeMessage assembly

0xfuturistic committed:
op-chain-ops: fix layer_two_test

0xfuturistic committed:
op-node: update L1Block bin in ecotone_upgrade_transactions

0xfuturistic committed:
contracts-bedrock: fix Initializable.t.sol

0xfuturistic committed:
contracts-bedrock: update snapshots

0xfuturistic committed:
contracts-bedrock: add semver to CrossL2Inbox, L2ToL2CrossDomainMessenger

0xfuturistic committed:
op-bindings: make

tynes commented on packages/contracts-bedrock/src/libraries/SafeCall.sol:
```diff
@@ -48,6 +48,27 @@ library SafeCall {
+    /// @param _target   Address to call
+    /// @param _value    Amount of value to pass to the call
+    /// @param _calldata Calldata to pass to the call
+    function callWithAllGas(address _target, uint256 _value, bytes memory _calldata) internal returns (bool) {
```
I don't think we can safely change the solidity version in this file unfortunately, perhaps we just put this function into the contract its used in

tynes commented on packages/contracts-bedrock/test/L2/L1Block.t.sol:
```diff
@@ -57,7 +65,9 @@ contract L1BlockBedrock_Test is L1BlockTest {
-            _l1FeeScalar: type(uint256).max
+            _l1FeeScalar: type(uint256).max,
+            _interopSetSize: uint8(_chainIds.length),
+            _chainIds: _chainIds
```
Thoughts on calling this "dependency set" to match the spec?

0xfuturistic replied:
I'm using `_chainIds` for consistency with https://github.com/ethereum-optimism/specs/blob/main/specs/interop/predeploys.md#l1attributes (check the name of the last input arg in L1Attributes's table). However, I think "dependency set" would be more descriptive. In that case, we should also update the table so that the arg is something like `dependencySet`, right? or even `interopSet`, so that it relates to `interopSetSize` more directly. alternatively, we could update `interopSetSize` to `dependencySetSize`. thoughts? cc @protolambda

0xfuturistic replied:
dropped `_interopSetSize` and changed `_chainIds` for `_dependencySet`

tynes commented on packages/sdk/src/utils/contracts.ts:
```diff
@@ -103,6 +105,10 @@ const getContractInterfaceBedrock = (name: string): ethers.utils.Interface => {
     case 'OptimismPortal2':
       artifact = optimismPortal2
       break
+    case 'CrossL2Inbox':
+      artifact = crossL2Inbox
+    case 'L2ToL2CrossDomainMessenger':
+      artifact = l2ToL2CrossDomainMessenger
```
This does need a single `break` statement

tynes replied:
The rabbit is right here

tynes commented on packages/contracts-bedrock/src/libraries/Predeploys.sol:
```diff
@@ -1,6 +1,75 @@
 // SPDX-License-Identifier: MIT
 pragma solidity ^0.8.0;
 
+interface ICrossL2Inbox {
```
Are these interfaces used anywhere? We don't maintain interfaces for the other predeploys

0xfuturistic replied:
unfortunately we use interfaces since `CrossL2Inbox.sol` and `L2ToL2CrossDomainMessenger.sol` are `^0.8.24` whereas setup files and tests are `0.8.15`, and can't import `^0.8.24` contracts directly. To not have to change the version of setup files and tests, I used interfaces for the `^0.8.24` contracts. assuming that's a reasonable solution, where does it make sense to declare them?

0xfuturistic replied:
ended up creating an interface file for each 0.8.24 contract but in 0.8.15 and stored them in src/L2

tynes commented on packages/contracts-bedrock/src/L2/L2ToL2CrossDomainMessenger.sol:
```diff
@@ -0,0 +1,135 @@
+    /// @param _value       ETH value to send with the message.
+    /// @param _message     Message to send to the target.
+    function relayMessage(
+        uint256 _destination,
```
This needs to be updated to include https://github.com/ethereum-optimism/specs/pull/81

tynes commented on packages/contracts-bedrock/src/L2/L2ToL2CrossDomainMessenger.sol:
```diff
@@ -0,0 +1,135 @@
+    uint16 public constant MESSAGE_VERSION = uint16(0);
+
+    /// @notice Initial balance for the contract.
+    uint248 public constant INITIAL_BALANCE = type(uint248).max;
```
We don't need initial balance anymore, it has been removed from the spec

tynes commented on packages/contracts-bedrock/src/L2/L2ToL2CrossDomainMessenger.sol:
```diff
@@ -0,0 +1,135 @@
+        require(successfulMessages[messageHash] == false, "L2ToL2CrossDomainMessenger: message already relayed");
+
+        assembly {
+            tstore(CROSS_DOMAIN_MESSAGE_SENDER_SLOT, _sender)
```
We need to make sure to have a getter for this

tynes commented on packages/contracts-bedrock/src/L2/L2ToL2CrossDomainMessenger.sol:
```diff
@@ -0,0 +1,135 @@
+        if (success) {
+            // This check is identical to one above, but it ensures that the same message cannot be relayed
+            // twice, and adds a layer of protection against rentrancy.
+            assert(successfulMessages[messageHash] == false);
```
Reentrency is impossible given that this can only be called by the cross l2 inbox which has 1 function that is only eoa

tynes commented on packages/contracts-bedrock/src/L2/L2ToL2CrossDomainMessenger.sol:
```diff
@@ -0,0 +1,135 @@
+        uint256 _nonce,
+        address _sender,
+        address _target,
+        uint256 _value,
```
value has been removed from the spec

tynes commented on packages/contracts-bedrock/src/L2/L2ToL2CrossDomainMessenger.sol:
```diff
@@ -0,0 +1,135 @@
+            tstore(CROSS_DOMAIN_MESSAGE_SENDER_SLOT, _sender)
+        }
+
+        bool success = SafeCall.callWithAllGas({ _target: _target, _value: _value, _calldata: _message });
```
this should send `msg.value`

tynes commented on packages/contracts-bedrock/src/L2/L2ToL2CrossDomainMessenger.sol:
```diff
@@ -0,0 +1,135 @@
+            L2ToL2CrossDomainMessenger.relayMessage,
+            (_destination, messageNonce(), msg.sender, _target, msg.value, _message)
+        );
+        emit SentMessage(_destination, _target, _message, data);
```
This doesn't follow the spec, it needs to only be the data to work

tynes commented on packages/contracts-bedrock/src/L2/L2ToL2CrossDomainMessenger.sol:
```diff
@@ -0,0 +1,135 @@
+    uint248 public constant INITIAL_BALANCE = type(uint248).max;
+
+    /// @notice Address of the L2 Cross Domain Messenger on this chain.
+    address public crossL2Inbox;
```
This is never set to anything, we don't need this

0xfuturistic replied:
are you sure we don't need it? `crossL2Inbox` is used in the `relayMessage` method to check that the sender is `CrossL2Inbox` and that the origin in `CrossL2Inbox` is this contract. the value for `crossL2Inbox` is set through l2 genesis (which is why in the contract is not set to anything. maybe we could add a constructor to `L2ToL2CrossDomainMessenger` that additionally sets it to `Predeploys.CROSS_L2_INBOX`? this would be unnecessary though, since the value is already set

tynes commented on packages/contracts-bedrock/src/L2/L2ToL2CrossDomainMessenger.sol:
```diff
@@ -0,0 +1,135 @@
+    {
+        require(msg.sender == crossL2Inbox, "L2ToL2CrossDomainMessenger: sender not CrossL2Inbox");
+        require(
+            CrossL2Inbox(crossL2Inbox).origin() == address(this),
```
We should inline `Predeploys.CrossL2Inbox` here

tynes commented on packages/contracts-bedrock/src/L2/L2ToL2CrossDomainMessenger.sol:
```diff
@@ -0,0 +1,135 @@
+            successfulMessages[messageHash] = true;
+            emit RelayedMessage(messageHash);
+        } else {
+            failedMessages[messageHash] = true;
```
We don't need `failedMessages` for anything, just the event is fine

tynes commented on packages/contracts-bedrock/src/L2/L1Block.sol:
```diff
@@ -49,6 +49,12 @@ contract L1Block is ISemver {
+    uint8 public interopSetSize;
+
+    /// @notice The chain IDs in the interop dependency set.
+    uint256[] public chainIds;
```
can we rename this to be `dependencySet` to match the spec?

0xfuturistic replied:
see https://github.com/ethereum-optimism/optimism/pull/9690#discussion_r1516722809

tynes replied:
Ah good catch, we should certainly update the spec to use "dependency set" as much as possible

tynes commented on packages/contracts-bedrock/src/L2/L1Block.sol:
```diff
@@ -49,6 +49,12 @@ contract L1Block is ISemver {
+    uint256[] public chainIds;
+
     /// @custom:semver 1.2.0
     string public constant version = "1.2.0";
```
semver needs a minor bump

tynes commented on packages/contracts-bedrock/src/L2/CrossL2Inbox.sol:
```diff
@@ -0,0 +1,95 @@
+    /// @param _msg The message payload, matching the initiating message.
+    /// @param _id A Identifier pointing to the initiating message.
+    /// @param _target Account that is called with _msg.
+    function executeMessage(Identifier calldata _id, address _target, bytes calldata _msg) public payable {
```
The calldata ordering doesn't match the spec, any reason for that?

0xfuturistic replied:
included `_id` first because we use `calldataload` in assembly for getting its fields, and by including `_id` first, we don't need to figure out the exact location of `_id` within the calldata (especially important given that `_msg` can be of arbitrary size). thoughts?

tynes commented on packages/contracts-bedrock/src/L2/CrossL2Inbox.sol:
```diff
@@ -0,0 +1,95 @@
+    function executeMessage(Identifier calldata _id, address _target, bytes calldata _msg) public payable {
+        require(_id.timestamp <= block.timestamp, "CrossL2Inbox: invalid id timestamp"); // timestamp invariant
+        uint256 chainId_ = _id.chainId;
+        require(L1Block(l1Block).isInDependencySet(chainId_), "CrossL2Inbox: invalid id chainId"); // chainId invariant
```
We can inline `Predeploys.L1_BLOCK`

tynes commented on packages/contracts-bedrock/src/L2/CrossL2Inbox.sol:
```diff
@@ -0,0 +1,95 @@
+    // bytes32(uint256(keccak256("crossl2inbox.identifier.chainid")) - 1)
+    bytes32 public constant CHAINID_SLOT = 0x6e0446e8b5098b8c8193f964f1b567ec3a2bdaeba33d36acb85c1f1d3f92d313;
+
+    address public l1Block;
```
No need for this variable

tynes commented on packages/contracts-bedrock/src/L2/L1Block.sol:
```diff
@@ -116,4 +128,68 @@ contract L1Block is ISemver {
+                mstore(0x00, 0x3cc50b45) // 0x3cc50b45 is the 4-byte selector of "NotDepositor()"
+                revert(0x1C, 0x04) // returns the stored 4-byte selector from above
+            }
+
```
This function also needs to set the other values in storage besides just the dependency set

0xfuturistic replied:
sounds good, but just to confirm: given the fact that ["[the] interop upgrade block itself MUST include a call to setL1BlockValuesEcotone"](https://github.com/ethereum-optimism/specs/blob/main/specs/interop/predeploys.md#l1attributes:~:text=The%20interop%20upgrade%20block%20itself%20MUST%20include%20a%20call%20to%20setL1BlockValuesEcotone.) the other storage values are written by `setL1BlockValuesEcotone` for every time `setL1BlockValuesInterop` is called. thus, isn't it unnecessary to rewrite those values by `setL1BlockValuesInterop`?

tynes replied:
I don't follow exactly - the idea is that the first tx in every block is always the L1 attributes tx, we cannot break that invariant. The upgrade txs will deploy + upgrade the L1Block predeploy afterwards meaning that the interop interface isn't available to be called until interop activation height + 1

The smart contract parses the calldata built in Go when creating the L1 attributes tx, it still passes the rest of the values from the ecotone upgrade, the interop upgrade extends the ecotone upgrade

0xfuturistic replied:
aah ok, so `setL1BlockValuesInterop` replaces `setL1BlockValuesEcotone`, right?

tynes commented on packages/contracts-bedrock/src/L2/L1Block.sol:
```diff
@@ -116,4 +128,68 @@ contract L1Block is ISemver {
+    ///   7. _blobBaseFee        L1 blob base fee.
+    ///   8. _hash               L1 blockhash.
+    ///   9. _batcherHash        Versioned hash to authenticate batcher by.
+    ///  10. _interopSetSize     Size of the interop dependency set.
```
nit: dependency set

0xfuturistic replied:
see https://github.com/ethereum-optimism/optimism/pull/9690#discussion_r1516722809

tynes commented on packages/contracts-bedrock/src/L2/L1Block.sol:
```diff
@@ -1,5 +1,5 @@
 // SPDX-License-Identifier: MIT
-pragma solidity 0.8.15;
+pragma solidity ^0.8.15;
```
must this semver be bumped?

0xfuturistic committed:
contracts-bedrock: fix pragma of dependencies of CrossL2Inbox

0xfuturistic committed:
contracts-bedrock: fix pragma of dependencies of L2ToL2CrossDomainMessenger

0xfuturistic committed:
contracts-bedrock: refactor callWithAllGas references

0xfuturistic committed:
contracts-bedrock: redefine assembly for CrossL2Inbox

0xfuturistic committed:
contracts-bedrock: add missing var in L2ToL2CrossDomainMessenger

0xfuturistic committed:
contracts-bedrock: add revert log for CrossL2Inbox

0xfuturistic committed:
contracts-bedrock: fix CrossL2Inbox assembly

0xfuturistic committed:
contracts-bedrock: drop unnecessary variable in CrossL2Inbox

0xfuturistic committed:
contracts-bedrock: minor bump semver for L1Block

0xfuturistic committed:
contracts-bedrock: drop unnecessary variable in L2ToL2CrossDomainMessenger

0xfuturistic committed:
contracts-bedrock: drop failedMessages in L2ToL2CrossDomainMessenger

0xfuturistic committed:
contracts-bedrock: drop unnecessary reentrancy check in L2ToL2CrossDomainMessenger

0xfuturistic committed:
contracts-bedrock: drop unnecessary arg in L2ToL2CrossDomainMessenger

0xfuturistic committed:
contracts-bedrock: drop unnecessary var in L2ToL2CrossDomainMessenger

0xfuturistic committed:
contracts-bedrock: add getter L2ToL2CrossDomainMessenger

0xfuturistic committed:
contracts-bedrock: fix data encoding in L2ToL2CrossDomainMessenger

0xfuturistic committed:
contracts-bedrock: add source to L2ToL2CrossDomainMessenger

0xfuturistic committed:
contracts-bedrock: fix after changes to L2ToL2CrossDomainMessenger, CrossL2Inbox

0xfuturistic committed:
op-bindings: make

0xfuturistic committed:
op-chain-ops: update storage variables for CrossL2Inbox, L2ToL2CrossDomainMessenger

0xfuturistic committed:
contracts-bedrock: update error message for CrossL2Inbox

0xfuturistic committed:
contracts-bedrock: fix test for CrossL2Inbox

0xfuturistic committed:
contracts-bedrock: drop unnecessary test for L2ToL2CrossDomainMessenger

0xfuturistic committed:
op-bindings: make

0xfuturistic committed:
contracts-bedrock: update tests for L2ToL2CrossDomainMessenger

0xfuturistic committed:
op-node: fix tests

0xfuturistic committed:
sdk: fix L1Block import

0xfuturistic committed:
contracts-bedrock: rename _tstoreCalldataID to _tStoreCalldataID in CrossL2Inbox

0xfuturistic committed:
contracts-bedrock: refactor comments for L1Block.sol

0xfuturistic committed:
contracts-bedrock: refactor comments for CrossL2Inbox.sol

0xfuturistic committed:
contracts-bedrock: refactor comments for L2ToL2CrossDomainMessenger.sol

0xfuturistic committed:
contracts-bedrock: refactor CrossL2Inbox

0xfuturistic committed:
contracts-bedrock: create CrossL2Inbox interface

0xfuturistic committed:
contracts-bedrock: create L2ToL2CrossDomainMessenger interface

0xfuturistic committed:
contracts-bedrock: refactor L2ToL2CrossDomainMessenger

0xfuturistic committed:
contracts-bedrock: remove interfaces from Predeploys

0xfuturistic committed:
contracts-bedrock: drop redundant tests for L2ToL2CrossDomainMessenger

0xfuturistic committed:
op-node: add support for interop in L1Block

0xfuturistic committed:
op-bindings: make

0xfuturistic committed:
op-node: fix fuzz_parsers_test

0xfuturistic committed:
op-node: create FuzzL1InfoInteropRoundTrip

0xfuturistic committed:
op-node: fix length of L1Info

0xfuturistic committed:
op-node: fix l1_block_info interop

0xfuturistic committed:
op-service: drop unused util methods

0xfuturistic committed:
op-node: fix fuzz_parsers_test

0xfuturistic committed:
contracts-bedrock: refactor tests for L2ToL2CrossDomainMessenger

tynes commented on op-node/rollup/derive/l1_block_info.go:
```diff
@@ -18,8 +18,9 @@ import (
 
 const (
-	L1InfoFuncBedrockSignature = "setL1BlockValues(uint64,uint64,uint256,bytes32,uint64,bytes32,uint256,uint256)"
+	L1InfoFuncBedrockSignature = "setL1BlockValues(uint64,uint64,uint256,bytes32,uint64,bytes32,uint256,uint256,uint8,uint256[])"
```
We cannot change this - it would be a hardfork

tynes commented on op-node/rollup/derive/l1_block_info.go:
```diff
@@ -54,6 +61,9 @@ type L1BlockInfo struct {
 	BaseFeeScalar     uint32   // added by Ecotone upgrade
 	BlobBaseFeeScalar uint32   // added by Ecotone upgrade
+
+	InteropSetSize uint8      // added by Interop upgrade
```
The size is implicit in `len(ChainIds)` so i don't think we need an explicit field for it

tynes commented on op-service/eth/types.go:
```diff
@@ -365,6 +365,10 @@ type SystemConfig struct {
+	// InteropSetSize identifies the size of the interop set
+	InteropSetSize uint8 `json:"interopSetSize"`
+	// ChainIds identifies the chain IDs of the chains in the interop depenency set
+	ChainIds []*big.Int `json:"chainIds"`
```
Can we rename this to be "dependencySet` everywhere so it matches the spec 1:1? We should also be able to get rid of `InteropSetSize` since its the same as len(dependencySet)`

tynes commented on op-service/eth/types.go:
```diff
@@ -418,6 +422,28 @@ func CheckEcotoneL1SystemConfigScalar(scalar [32]byte) error {
 	}
 }
 
+func (sysCfg *SystemConfig) InteropScalars() (blobBaseFeeScalar, baseFeeScalar uint32, err error) {
```
How is this different from ecotone? If we need this it should just be able to wrap `EcotoneScalars` similar to in geth the `Signer` class is updated for each hardfork

0xfuturistic replied:
deleted the definition for `InteropScalars` since we use `EcotoneScalars` directly in it's place

tynes commented on packages/contracts-bedrock/src/L2/CrossL2Inbox.sol:
```diff
@@ -0,0 +1,139 @@
+
+        // Call the target account with the message payload.
+        bool success;
+        assembly {
```
Slight preference for wrapping this assembly into an internal function

tynes commented on packages/contracts-bedrock/src/L2/CrossL2Inbox.sol:
```diff
@@ -0,0 +1,139 @@
+
+/// @title IL1Block_OnlyIsInDependencySet
+/// @notice Interface for L1Block with only `isInDependencySet(uint256)` method.
+interface IL1Block_OnlyIsInDependencySet {
```
This name is a bit verbose, what do you think about something smaller like `IDependencySet` or something like that?

0xfuturistic replied:
great idea! just changed it

tynes commented on versions.json:
```diff
@@ -3,6 +3,6 @@
   "geth": "v1.13.4",
   "nvm": "v20.9.0",
-  "slither": "0.10.0",
+  "slither": "0.10.1",
```
thoughts on breaking this out into its own PR?

0xfuturistic replied:
rebased and looks slither-analyze now uses 0.10.1, so this change is now unnecessary
